### PR TITLE
Refactor attributes matching API

### PR DESCRIPTION
**Pull request by melix**
This pull request refactors the attributes matching API (see gradle/performance#237) and is a pre-requisite to gradle/performance#233.

If there are no major issues with this, I think we should merge asap to avoid too many merge conflicts.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/attributes/MultipleCandidatesDetails.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * @param <T> the concrete type of the attribute
+ */
+@Incubating
+public interface MultipleCandidatesDetails<T> {
+    /**
+     * The value of the attribute on the consumer side. All possible outcomes are valid (present, missing, unknown)
+     * @return the value of the attribute as found on the consumer side
+     */
+    AttributeValue<T> getConsumerValue();
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/api/attributes/MultipleCandidatesDetails.java]: Perhaps just `T` and we can deal with missing values differently.

File: subprojects/core/src/main/java/org/gradle/api/attributes/MultipleCandidatesDetails.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * A map of candidate values. The keys represent a candidate, and the value is the value of the attribute for the attribute.
+     * @return the map of candidates
+     */
+    Map<HasAttributes, AttributeValue<T>> getCandidateValues();
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/api/attributes/MultipleCandidatesDetails.java]: I'd simplify this by returning `Set<AttributeValue<T>>` or even better `Set<T>`. The rule doesn't really care about the keys of this map and has to do a reverse lookup once it selects the best value in order to provide one of the keys to `closestMatch()`.

File: subprojects/core/src/main/java/org/gradle/api/attributes/MultipleCandidatesDetails.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * Calling this method indicates that the candidate is the closest match. It is allowed to call this method several times with
+     * different keys, in which case it indicates that multiple candidates are equally compatible.
+     * @param key the candidate which is the closest match
+     */
+    void closestMatch(HasAttributes key);
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/api/attributes/MultipleCandidatesDetails.java]: I'd make the parameter `AttributeValue<T>` or `T`

File: subprojects/core/src/main/java/org/gradle/api/attributes/OrderedCompatibilityRule.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+/**
+ * <p>An ordered compatibility rule will use the provided {@link java.util.Comparator} to compare values. It will
+ * only take a decision if both the consumer and producer attributes are present.</p>
+ *
+ * <p>A provider attribute is compatible with the consumer attribute if it's lower than or equal to the consumer value.</p>
+ *
+ * @param <T>
+ */
+@Incubating
+public interface OrderedCompatibilityRule<T> extends CompatibilityRule<T> {
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/api/attributes/OrderedCompatibilityRule.java]: do we need this type in the public API? What would the client do with this?

File: subprojects/core/src/main/java/org/gradle/api/attributes/OrderedDisambiguationRule.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ */
+@Incubating
+public interface OrderedDisambiguationRule<T> extends DisambiguationRule<T> {
+    /**
+     * Disambiguate by selecting the first candidate in order. If the attribute value
+     * is missing or unknown, then it is excluded from the list of candidate.
+     *
+     * @return this rule
+     */
+    OrderedDisambiguationRule<T> pickFirst();
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/api/attributes/OrderedDisambiguationRule.java]: I'd just make these overloads on `DisambiguationRuleChain`, something like `pickFirst(Comparator<T>)` and `pickLast(Comparator<T>)`.

File: subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultAttributesSchema.java [hunk 6]
```diff
@@ -16,63 +16,49 @@
...
         if (strategy == null) {
             throw new IllegalArgumentException("Unable to find matching strategy for " + attribute);
         }
         return Cast.uncheckedCast(strategy);
     }
 
     @Override
-    public <T> void matchStrictly(Attribute<T> attribute) {
-        setMatchingStrategy(attribute, StrictMatchingStrategy.<T>get());
+    public <T> AttributeMatchingStrategy<T> configureMatchingStrategy(Attribute<T> attribute, Action<? super AttributeMatchingStrategy<T>> configureAction) {
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultAttributesSchema.java]: It looks like someone has to explicitly define each attribute. Possibly we should just default to equality checking for attributes we don't know about.

File: subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java [hunk 7]
```diff
@@ -120,11 +122,31 @@
...
 import static java.util.Collections.singletonMap;
+import static org.gradle.api.internal.artifacts.ArtifactAttributes.ARTIFACT_CLASSIFIER;
+import static org.gradle.api.internal.artifacts.ArtifactAttributes.ARTIFACT_EXTENSION;
+import static org.gradle.api.internal.artifacts.ArtifactAttributes.ARTIFACT_FORMAT;
 import static org.gradle.util.GUtil.addMaps;
 import static org.gradle.util.GUtil.isTrue;
 
 @NoConventionMapping
 public class DefaultProject extends AbstractPluginAware implements ProjectInternal, DynamicObjectAware {
+    private static final Action<AttributeMatchingStrategy<String>> ARTIFACT_ATTRIBUTE_CONFIG = new Action<AttributeMatchingStrategy<String>>() {
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java]: This stuff doesn't really belong in project. We should move the `AttributeSchema` to live on `DependencyHandler` with the other dependency resolution rules/logic

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ArtifactSelectionIntegrationTest.groovy [hunk 8]
```diff
@@ -36,6 +36,11 @@ allprojects {
         repositories {
             ivy { url '${ivyHttpRepo.uri}' }
         }
+        configurationAttributesSchema {
```
Comment (adammurdoch) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ArtifactSelectionIntegrationTest.groovy]: this should just be the default.

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/resolveengine/result/StreamingResolutionMissingOrUnknownHandlerTest.groovy [hunk 9]
```diff
@@ -29,7 +29,7 @@ import spock.lang.Specification
 import static org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.ResolutionResultPrinter.printGraph
 import static org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.VersionSelectionReasons.*
 
-class StreamingResolutionResultBuilderTest extends Specification {
+class StreamingResolutionMissingOrUnknownHandlerTest extends Specification {
```
Comment (melix) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/resolveengine/result/StreamingResolutionMissingOrUnknownHandlerTest.groovy]: Oops!

File: subprojects/core/src/main/java/org/gradle/api/attributes/CompatibilityRuleChain.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    /**
+     * Adds an ordered check rule to this chain. It is expected to call this method if and only if the
+     * type of the attribute is {@link Comparable}. In that case compatibility is working as described
+     * in {@link OrderedCompatibilityRule}.
+     *
+     * @param comparator the comparator to use
+     *
+     * @return the added ordered check rule
+     */
+    OrderedCompatibilityRule<T> addOrderedCheck(Comparator<? super T> comparator);
```
Comment (melix) [inline subprojects/core/src/main/java/org/gradle/api/attributes/CompatibilityRuleChain.java]: Funnily I had started with the `T extends Comparable<T>` version, but it turned out to be a mess due to generics (type safety gone, thank you Java). See https://github.com/gradle/gradle/pull/953/commits/fdf58d74f1db70a839c29b575b88e296fd00226c

File: subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultAttributesSchema.java (hunk 6 above)
Comment (melix) [inline subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultAttributesSchema.java]: Yes, the change was required in order to be able to unify how we handle full and partial matches. See the discussion at https://gradle.slack.com/archives/performance/p1480518840000719

File: subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java (hunk 7 above)
Comment (melix) [inline subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java]: Agreed, I didn't want to tackle this in this story (to be able to do gradle/performance#233 first). I'd like to have something merged before shuffling things around.

**Comment by melix** [comment]
All changes implemented, except for implementing rules for "present only" values. @oehme and I had a long discussion about this this morning, and it appears we would better move the schema to `gradle` itself (global scope) and deal with unknown and missing attributes consistently. It doesn't say anything about the requirement to have access to the presence or absence of attributes at the rule level, instead of chain level, but it would simplify the implementation and some awkwardness.

**Pull request merged**